### PR TITLE
[FrameIt] Multiple layout improvements (including iPhone X, multiline labels)

### PR DESCRIPTION
**Pull request by bvirlet**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

_frameit_ has some limitations to its layout engine. With new devices such as the iPhone X, this sometimes results in awkward results. This PR addresses a couple of these layout issues and adds options to customize the layout. This are all useful changes that I applied for the latest version of our app Genius Scan.

### Description

This addresses multiple layout issues:

#### Multi-lines labels were improperly laid out

A multiline label is a string with a `\n`.

beforeafter

Note that the layout isn't particularly "beautiful" but I have set the padding to 0 to be clear about the consequences. With a large padding, the problem was less visible.

This is fixed by not relying on the font size anymore (it would always return the height of 1 line) but on the actual title height which returns the height for the multiple lines.

#### Add title_min_height param to avoid iPhone X layout issues

This enables to have the device tops being all at the same height even if the title/keyword are taller on a given screenshot. This can happen with multiline titles but also titles were the size was adjusted (see example). **Having devices at different heights is not super nice when the user scrolls horizontally between the screenshots on the App Store.**

Additionally, it's another way to add some breathing room to the screenshots, independently from the `padding` option.

_Before_

_After_, with `"title_min_height" : "20%"`

#### Add cascading filters

You may want to adjust the layout not only based on the screenshot number/name, but also on the device. Currently, only the first filter matching the screenshot path is applied. With this change, the filters are applied in order.

This lets modify the iPad screenshot without modifying the iPhone's:

```
{
  "device_frame_version": "latest",
  "default": {
    "title": {
      "color": "#ffffff"
    },
    "background": "./Frameit-background.png",
    "padding": "3%",
    "show_complete_frame": false,
    "stack_title" : true,
    "title_below_image": false,
    "font_scale_factor": 0.05
  },
   "data": [
    {
      "filter": "scan"
    },
    {
      "filter": "edit"
    },
    {
      "filter": "multipage_pdf"
    },
    {
      "filter": "organize"
    },
    {
      "filter": "export"
    },
    {
      "filter" : "iPad",
      "show_complete_frame" : true,
      "font_scale_factor": 0.035
    }
  ]
}
```

_Before_
before

_After_
after

Another example is how this lets us improve the iPhone X screenshots together with the `title_min_height` feature: 

```
    {
      "filter" : "iPhone X",
      "title_min_height" : "30%"
    }
```

_Before_

 

_After_

### CODE REVIEW

**Comment by Shebuka** [comment]
This will be so useful to see merged.

**Comment by dayjer** [comment]
Would be great to see this reviewed so it can be merged.

File: fastlane/lib/fastlane/actions/docs/frameit.md [hunk 1]
```diff
@@ -131,7 +131,9 @@ The `show_complete_frame` value specifies whether _frameit_ should shrink the de
 
 The `title_below_image` value specifies whether _frameit_ should place the title below the screenshot. If it is false, it will be placed above the screenshot.
 
-The `filter` value is a part of the screenshot named for which the given option should be used. If a screenshot is named `iPhone5_Brainstorming.png` the first entry in the `data` array will be used.
+The `title_min_height` value specifies whether _frameit_ should keep a minimum height for the text area, even if the text's actual height is smaller.
```
Comment (janpio) [inline fastlane/lib/fastlane/actions/docs/frameit.md]: What type of value does this expect? boolean, px, %?

**Comment by janpio** [comment]
This PR seems pretty useful, but is not the easiest to review as it contains 3 different changes. At least for me it is too complicated right now. I could take another look if @bvirlet maybe moves feature 2 and 3 into their own branch and PRs? That would help a lot.

**Comment by bvirlet** [comment]
Hey Jan. I will see if I can find the time to do that. The difficulty is that the changes are not so independent, so it's a bit hard to split them. I understand why this would be helpful, of course.

**Comment by dayjer** [comment]
Any updates on this? My iPhone X screenshots look horrendous :)

**Comment by bvirlet** [comment]
At the moment I don't have the time to split the PR. The difficulty is that these changes are a bit interdependent so it will basically require me to rewrite everything. @dayjer if you're in a hurry you can use the development version of fastlane and checkout this branch.

**Comment by janpio** [comment]
Damn, unfortunately this PR has gone stale because of changes in `master`, the conflicts can't just be resolved in the Github UI so you @bvirlet would have to rebase onto `master` locally and then push this up again.

**Comment by bvirlet** [comment]
Hi @janpio : I can do this but is it with the goal of reviewing it and merging it afterwards "as is" ?

**Comment by janpio** [comment]
Let's see about that - I will pull in some additional people and/or maybe try to extract parts of it into their own PR. But the end goal is to get this merged, yes.

**Comment by joshdholtz** [comment]
@futuretap Hey sorry! This PR got lost in the shuffle 😔 Can you rebase from master since it looks like there are some conflicts?

**Comment by futuretap** [comment]
@joshdholtz I didn't write that PR, so I'm not familiar with the code.

**Comment by futuretap** [comment]
I've created https://github.com/fastlane/fastlane/pull/12484 and rebased this PR from master. Feel free to pull it. (And no, I don't have to time to make rubocop happy.)

**Comment by bvirlet** [comment]
For reference I have rebased this changes on top of master here: https://github.com/thegrizzlylabs/fastlane/tree/layout-improvements

Unfortunately the PR #11585 only partially addresses the issues that #11014 solves but conflicts with it.

I think the solution for _frameit_ would be to have a LayoutManager class that would create a layout object and we could then have an extensive suite of tests on this layout object, avoiding regressions:
```
class Layout
  def title_rect; end
  def keyword_rect; end
  def …
end
```

I don't have the time right now to do this but maybe will find some in the future.

**Comment by getaaron** [comment]
Closing this and continuing conversation on https://github.com/fastlane/fastlane/pull/12484, which doesn't have conflicts

**Pull request closed without merging**